src/runtime/TemporalObject.cpp

```diff
@@ -1196,6 +1326,141 @@ std::map<std::string, int> TemporalPlainYearMonth::balanceISOYearMonth(Execution
+{
+    int sign = TemporalDuration::durationSign(fields);
+    for (int field : fields) {
+        if (field < 0 && sign > 0) {
+            return false;
+        } else if (field > 0 && sign < 0) {
+            return false;
+        }
+    }
+    return true;
+}
+
+int TemporalDuration::durationSign(const std::array<int, 10>& fields)
+{
+    for (int field : fields) {
+        if (field < 0) {
+            return -1;
+        } else if (field > 0) {
+            return 1;
+        }
+    }
+    return 0;
+}
+
+Value TemporalDuration::createTemporalDuration(ExecutionState& state, int years, int months, int weeks, int days, int hours, int minutes, int seconds, int milliseconds, int microseconds, int nanoseconds, Optional<Object*> newTarget)
+{
+    if (isValidDuration({ years, months, weeks, days, hours, minutes, seconds, milliseconds, microseconds, nanoseconds })) {
+        ErrorObject::throwBuiltinError(state, ErrorObject::RangeError, "Invalid time");
+    }
+
+    if (!newTarget.hasValue()) {
+        newTarget = state.resolveCallee();
+    }
+
+    Object* proto = Object::getPrototypeFromConstructor(state, newTarget.value(), [](ExecutionState& state, Context* constructorRealm) -> Object* {
+        return constructorRealm->globalObject()->temporalDurationPrototype();
+    });
+
+    return new TemporalDuration(state, proto, years, months, weeks, days, hours, minutes, seconds, milliseconds, microseconds, nanoseconds);
+}
+Value TemporalDuration::toTemporalDuration(ExecutionState& state, const Value& item)
+{
+    if (item.isObject() && item.asObject()->asTemporalDurationObject()) {
+        return item;
+    }
+
+    auto result = TemporalDuration::toTemporalDurationRecord(state, item);
+
+    return TemporalDuration::createTemporalDuration(state, result["years"], result["months"], result["weeks"], result["days"], result["hours"], result["minutes"], result["seconds"], result["milliseconds"], result["microseconds"], result["nanoseconds"], nullptr);
+}
+
+std::map<std::string, int> TemporalDuration::toTemporalDurationRecord(ExecutionState& state, const Value& temporalDurationLike)
+{
+    if (!temporalDurationLike.isObject()) {
+        return TemporalObject::parseTemporalDurationString(state, temporalDurationLike.asString()->toNonGCUTF8StringData());
+    } else if (temporalDurationLike.asObject()->isTemporalDurationObject()) {
+        auto duration = temporalDurationLike.asObject()->asTemporalDurationObject();
+        return TemporalDuration::createDurationRecord(state, duration->getYear(), duration->getMonth(), duration->getWeek(), duration->getDay(), duration->getHour(), duration->getMinute(), duration->getSecond(), duration->getMillisecond(), duration->getMicrosecond(), duration->getNanosecond());
+    }
+
+    std::map<std::string, int> result = { { "years", 0 }, { "months", 0 }, { "weeks", 0 }, { "days", 0 }, { "hours", 0 }, { "minutes", 0 }, { "seconds", 0 }, { "milliseconds", 0 }, { "microseconds", 0 }, { "nanoseconds", 0 } };
```

Comment:
Is it possible to use `enum` instead of string?

Reply:
Yes but everywhere else I use strings.